src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp

```diff
@@ -986,20 +993,45 @@ namespace Opm {
 
                 properties.injectorType = injectorType;
 
-                const std::string& cmodeString = record.getItem("CMODE").getTrimmedString(0);
-                WellInjector::ControlModeEnum controlMode = WellInjector::ControlModeFromString( cmodeString );
                 if (!record.getItem("RATE").defaultApplied(0)) {
                     properties.surfaceInjectionRate = injectionRate;
-                    properties.addInjectionControl(controlMode);
-                    properties.controlMode = controlMode;
                 }
-                properties.predictionMode = false;
 
                 if ( record.getItem( "BHP" ).hasValue(0) )
                     properties.BHPH = record.getItem("BHP").getSIDouble(0);
                 if ( record.getItem( "THP" ).hasValue(0) )
                     properties.THPH = record.getItem("THP").getSIDouble(0);
 
+                const std::string& cmodeString = record.getItem("CMODE").getTrimmedString(0);
+                const WellInjector::ControlModeEnum controlMode = WellInjector::ControlModeFromString( cmodeString );
+
+                if ( !(controlMode == WellInjector::RATE || controlMode == WellInjector::BHP) ) {
+                    const std::string msg = "Only RATE and BHP control are allowed for WCONINJH for well " + well->name();
+                    throw std::invalid_argument(msg);
+                }
+
+                // when well is under BHP control, we use its historical BHP value as BHP limit
+                if (controlMode == WellInjector::BHP) {
+                    properties.BHPLimit = properties.BHPH;
+                } else {
+                    const bool switching_from_producer = well->isProducer(currentStep);
+                    const bool switching_from_prediction = properties.predictionMode;
+                    const bool switching_from_BHP_control = (properties.controlMode == WellInjector::BHP);
+                    if (switching_from_prediction ||
+                        switching_from_BHP_control ||
+                        switching_from_producer) {
+                        // we use defaulted BHP value, it is from simulation, without finding any related document
+                        properties.BHPLimit = 6891.2 * unit::barsa;
```

Comment:
Since the `BHPLimit` property is occluded by large amount of occult black magic I would prefer if it were hidden as a private member with methods:
```
double get_BHPLimit() const;
void set_BHPLImit(double limit);
void reset_default_BHPLimit();
```
Then the actual default values of `1 atm` and `6891.2 BARS` can be hidden away inside the implementation, and the method name `reset_default_BHPLimit()` will communicate quite clearly what actually happens.